src/particles/elements/SoftSol.H

```diff
@@ -0,0 +1,402 @@
+        ) const
+        {
+            using namespace amrex::literals; // for _rt and _prt
+
+            // access AoS data such as positions and cpu/id
+            amrex::ParticleReal const x = p.pos(0);
+            amrex::ParticleReal const y = p.pos(1);
+            amrex::ParticleReal const t = p.pos(2);
+
+            // initialize output values of momenta
+            amrex::ParticleReal pxout = px;
+            amrex::ParticleReal pyout = py;
+            amrex::ParticleReal ptout = pt;
+
+            // get the linear map
+            amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const R = refpart.map;
+
+            // symplectic linear map for a solenoid is computed using the
+            // Hamiltonian formalism as described in:
+            // https://uspas.fnal.gov/materials/09UNM/ComputationalMethods.pdf.
+            // R denotes the transfer matrix in the basis (x,px,y,py,t,pt),
+            // so that, e.g., R(3,4) = dyf/dpyi.
+
+            // push particles using the linear map
+            p.pos(0) = R(1,1)*x + R(1,2)*px + R(1,3)*y
+                     + R(1,4)*py + R(1,5)*t + R(1,6)*pt;
+            pxout = R(2,1)*x + R(2,2)*px + R(2,3)*y
+                  + R(2,4)*py + R(2,5)*t + R(2,6)*pt;
+            p.pos(1) = R(3,1)*x + R(3,2)*px + R(3,3)*y
+                     + R(3,4)*py + R(3,5)*t + R(3,6)*pt;
+            pyout = R(4,1)*x + R(4,2)*px + R(4,3)*y
+                  + R(4,4)*py + R(4,5)*t + R(4,6)*pt;
+            p.pos(2) = R(5,1)*x + R(5,2)*px + R(5,3)*y
+                     + R(5,4)*py + R(5,5)*t + R(5,6)*pt;
+            ptout = R(6,1)*x + R(6,2)*px + R(6,3)*y
+                  + R(6,4)*py + R(6,5)*t + R(6,6)*pt;
+
+            // assign updated momenta
+            px = pxout;
+            py = pyout;
+            pt = ptout;
+        }
+
+        /** This pushes the reference particle.
+         *
+         * @param[in,out] refpart reference particle
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void operator() (RefPart & AMREX_RESTRICT refpart) const
+        {
+            using namespace amrex::literals; // for _rt and _prt
+
+            // assign input reference particle values
+            amrex::ParticleReal const x = refpart.x;
+            amrex::ParticleReal const px = refpart.px;
+            amrex::ParticleReal const y = refpart.y;
+            amrex::ParticleReal const py = refpart.py;
+            amrex::ParticleReal const z = refpart.z;
+            amrex::ParticleReal const pz = refpart.pz;
+            amrex::ParticleReal const pt = refpart.pt;
+            amrex::ParticleReal const s = refpart.s;
+            amrex::ParticleReal const sedge = refpart.sedge;
+
+            // initialize linear map (deviation) values
+            for (int i=1; i<7; i++) {
+               for (int j=1; j<7; j++) {
+                  if (i == j)
+                      refpart.map(i, j) = 1.0_prt;
+                  else
+                      refpart.map(i, j) = 0.0_prt;
```

Comment:
Compilers can sometimes optimize such constructs that use the ternary (conditional) operator better:
```suggestion
                  auto const default_value = (i == j) ? 1.0_prt : 0.0_prt;
                  refpart.map(i, j) = default_value;
```
https://en.cppreference.com/w/cpp/language/operator_other